[WIP] doc: bluetooth: services: wifi_prov: document provisioning protocol (#21261)
nrfconnect/sdk-nrf: doc/nrf/libraries/bluetooth/services/wifi_prov.rst

   :depth: 2

This library implements a Bluetooth® GATT service for Wi-Fi® provisioning.

> It is an implementation of the server side of Wi-Fi provosioning protocol defined by Nordic Semiconductor.

Reviewer: ```suggestion
It is an implementation of the server side of the Wi-Fi provisioning protocol defined by Nordic Semiconductor.
```

   =================== ======================= ======================== =======================================================================================
   Name                Data type               Status                   Notes
   =================== ======================= ======================== =======================================================================================
   scan_record         message                 Optional                 The information of SSID found in scan action.
   state               enum                    Optional                 The state of connection action.
   reason              enum                    Optional                 The reason of connection failure.
   =================== ======================= ======================== =======================================================================================

These definitions are available as :file:`.proto` files under the path to the library.

Workflow
========

Multiple workflows are defined in the form of message exchange.

Determine provisioning protocol version
---------------------------------------

> In this workflow, the configurator requests the Version message by a transport-layer specific method, and the target sends the Version message over the transport layer.

Reviewer: ```suggestion
In this workflow, the configurator requests the Version message using a transport-layer specific method, and the target sends the Version message over the transport layer.
```

> Get WiFi status
> ---------------

Reviewer: ```suggestion
Get Wi-Fi status
----------------
```

> In this workflow, the configurator sends a Request message, in which the op_code is set as GET_STATUS, over the transport layer.

Reviewer: ```suggestion
In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``GET_STATUS``, over the transport layer.
```

> The target receives the message, retrieve required information, and set up a Response message, in which the request_op_code is GET_STATUS, the status indicates whether the operation is successful and possible failure reason, and device_status carries the status of WiFi if the operation is successful.

Reviewer: ```suggestion
The target receives the message,  retrieves the required information, and sets up a ``Response`` message.
In the ``Response`` message, the ``request_op_code`` is ``GET_STATUS``, the status indicates whether the operation is successful and any possible failure reason, and ``device_status`` carries the Wi-Fi status if the operation is successful.
```

Start SSID scan
---------------

> In this workflow, the configurator sends a Request message, in which the op_code is set as START_SCAN, over the transport layer.

Reviewer: ```suggestion
In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``START_SCAN``, over the transport layer.
```

> The target receives the message, trigger the scan using given parameters, and set up a Response message, in which the request_op_code is START_SCAN, and the status indicates whether the operation is successful and possible failure reason.

Reviewer: ```suggestion
The target receives the message, triggers the scan using the given parameters, and sets up a ``Response`` message.
In the ``Response`` message, the ``request_op_code`` is ``START_SCAN``, and the status indicates whether the operation is successful and any possible failure reason.
```

> When scan is finished, the target will set up a Result message for each SSID found during the scan, and the information is in the scan_record field.

Reviewer: ```suggestion
When the scan is finished, the target will set up a ``Result`` message for each SSID found during the scan, and the information will be in the ``scan_record`` field.
```

Stop SSID scan
--------------

> In this workflow, the configurator sends a Request message, in which the op_code is set as STOP_SCAN, over the transport layer.

Reviewer: ```suggestion
In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``STOP_SCAN``, over the transport layer.
```

> The target receives the message, stop WiFi scan, and set up a Response message, in which the request_op_code is STOP_SCAN, and the status indicates whether the operation is successful and possible failure reason.

Reviewer: ```suggestion
The target receives the message, stops the Wi-Fi scan, and sets up a ``Response`` message.
In the ``Response`` message, the ``request_op_code`` is ``STOP_SCAN``, and the status indicates whether the operation is successful and any possible failure reason.
```

Connect to AP
-------------

> In this workflow, the configurator sends a Request message, in which the op_code is set as SET_CONFIG, over the transport layer.

Reviewer: ```suggestion
In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``SET_CONFIG``, over the transport layer.
```

> The target receives the message, apply the WiFi connection information, and set up a Response message, in which the request_op_code is SET_CONFIG, and the status indicates whether the operation is successful and possible failure reason.

Reviewer: ```suggestion
The target receives the message, applies the Wi-Fi connection information, and sets up a ``Response message``.
In the ``Response`` message, the ``request_op_code`` is ``SET_CONFIG``, and the status indicates whether the operation is successful and any possible failure reason.
```

> When connection state changes or attempt fails, the target will set up a Result message, and the state field indicates the current state of the WiFi, and reason field indicates the failure reason.

Reviewer: ```suggestion
When the connection state changes or an attempt fails, the target will set up a ``Result`` message.
The state field indicates the current state of the Wi-Fi, and the reason field indicates the failure reason.
```

Disconnect from AP
------------------

> In this workflow, the configurator sends a Request message, in which the op_code is set as FORGET_CONFIG, over the transport layer. The target receives the message, trigger the WiFi disconnection, and set up a Response message, in which the request_op_code is FORGET_CONFIG, and the status indicates whether the operation is successful and possible failure reason.

Reviewer: ```suggestion
In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``FORGET_CONFIG``, over the transport layer.
The target receives the message, triggers the Wi-Fi disconnection, and sets up a ``Response`` message, in which the ``request_op_code`` is ``FORGET_CONFIG``, and the status indicates whether the operation is successful and the possible failure reason.
```

> When connection state changes or attempt fails, the target will set up a Result message, and the state field indicates the current state of the WiFi, and reason field indicates the failure reason.

Reviewer: ```suggestion
When the connection state changes or an attempt fails, the target will set up a ``Result`` message.
The state field indicates the current state of the Wi-Fi, and the reason field indicates the failure reason.
```

The service is divided into three parts:

> * GATT service interface: Defines GATT service and serves as the transport layer for the provisioning protocol.
> * Task and event handling component: Implements the provisioning protocol.

Reviewer: ```suggestion
* GATT service interface: Defines the GATT service and serves as the transport layer for the provisioning protocol.
* Task and event handling component: Implements the provisioning protocol.
```

* Configuration management component: Manages the provisioning data (in RAM and flash) accessed by multiple threads.

Provisioning protocol
*********************

The protocol is designed to enable the provisioning of Wi-Fi products from Nordic Semiconductor.
It is an application layer protocol, and the underlying transport layer can be arbitrary.
It uses `Protocol Buffers`_ as the serialization scheme, to achieve platform independence.

> It adopts client-server model.

Reviewer: ```suggestion
It adopts a client-server model.
```

> The target, the device having Wi-Fi products to be provisioned (for example, nRF5340 DK on nRF7002 DK), hosts the server.

Reviewer: ```suggestion
The target,  which is the device having Wi-Fi products to be provisioned (for example, nRF5340 DK on nRF7002 DK), hosts the server.
```

> The configurator, the device holds the information to be passed to and applied on the target (for example, a smartphone), acts as the client.

Reviewer: ```suggestion
The configurator, which is the device that holds the information to be passed to and applied on the target (for example, a smartphone), acts as the client.
```

The message is the basic unit for carrying information and is exchanged between the configurator and the target.

Message types
=============

The protocol defines four message types:

> * ``Info``: The message is to indicate the version of the provisioning protocol implemented on the target.

Reviewer: ```suggestion
* ``Info``: The message is to indicate the ``version`` of the provisioning protocol implemented on the target.
```

* ``Request``: The message is for the configurator to request a particular action on the target.

  The following table details the fields of the ``Request`` message.

   =================== ======================= ======================== =======================================================================================
   Name                Data type               Status                   Notes
   =================== ======================= ======================== =======================================================================================
   op_code             enum                    Optional                 The value represents the action to be taken by the target.
   scan_params         enum                    Optional                 The value represents the scan parameters used for a scan action.
   config              message                 Optional                 The value represents the Wi-Fi configuration used for a connection action.
   =================== ======================= ======================== =======================================================================================

>   The following table details the five possible values of ``op_code`` field.

Reviewer: ```suggestion
  The following table details the five possible values of the ``op_code`` field.
```

   =================== ================================================ =============================
   Operation code      Description                                      Notes
   =================== ================================================ =============================
   GET_STATUS          Get the status of the device, such as
                       provisioning state, Wi-Fi connection state,
                       and more
   START_SCAN          Trigger an access point (AP) scan                ``scan_params`` is optional
   STOP_SCAN           Stop AP scan
   SET_CONFIG          Store Wi-Fi configuration and connect to an AP   ``config`` is required
   FORGET_CONFIG       Erase the configuration and disconnect
   =================== ================================================ =============================

* ``Response``: The message is for the target to provide the feedback on the result of an action to the configurator.

>   The following table details the fields of ``Response`` message.

Reviewer: ```suggestion
  The following table details the fields of the ``Response`` message.
```

   =================== ======================= ======================== =======================================================================================
   Name                Data type               Status                   Notes
   =================== ======================= ======================== =======================================================================================
   request_op_code     enum                    Optional                 The action the message is to respond.
   status              enum                    Optional                 The result of the action.
   device_status       message                 Optional                 The status of the Wi-Fi on the target.
   =================== ======================= ======================== =======================================================================================

>   The status field can take one of following values:

Reviewer: ```suggestion
  The ``status`` field can take one of following values:
```

  * ``SUCCESS``: The operation is dispatched successfully.
  * ``INVALID_ARGUMENT``: The argument is invalid.
  * ``INVALID_PROTO``: The message cannot be encoded or decoded.
  * ``INTERNAL_ERROR``: The operation cannot be dispatched properly.

* ``Result``: The message is for the target to provide feedback on the Wi-Fi status to the configurator asynchronously.

>   The following table details the fields of ``Result`` message.

Reviewer: ```suggestion
  The following table details the fields of the ``Result`` message.
```

   =================== ======================= ======================== =======================================================================================
   Name                Data type               Status                   Notes
   =================== ======================= ======================== =======================================================================================
   scan_record         message                 Optional                 The information of the SSID found in the scan action.
   state               enum                    Optional                 The state of the connection action.
   reason              enum                    Optional                 The reason for the connection failure.
   =================== ======================= ======================== =======================================================================================

These definitions are available as :file:`.proto` files in the library path.

Workflow
========

Multiple workflows are defined in the form of message exchange.

Determine provisioning protocol version
---------------------------------------

> In this workflow, the configurator requests the ``Info`` message using a transport-layer specific method, and the target sends the ``Info`` message over the transport layer.

Reviewer: ```suggestion
In this workflow, the configurator requests the ``Info`` message using a transport layer-specific method, and the target sends the ``Info`` message over the transport layer.
```

Get Wi-Fi status
----------------

In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``GET_STATUS``, over the transport layer.
The target receives the message, retrieves the required information, and sets up a ``Response`` message.

> In the ``Response`` message, the ``request_op_code`` is ``GET_STATUS``, the ``status`` indicates whether the operation is successful and possible failure reason, and ``device_status`` carries the status of WiFi if the operation is successful.

Reviewer: ```suggestion
In the ``Response`` message, the ``request_op_code`` is ``GET_STATUS``, the ``status`` indicates whether the operation is successful and the possible failure reason, and ``device_status`` carries the Wi-Fi status if the operation is successful.
```

Start SSID scan
---------------

In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``START_SCAN``, over the transport layer.
The target receives the message, triggers the scan using the given parameters, and sets up a ``Response`` message.

> In the ``Response`` message, the ``request_op_code`` is ``START_SCAN``, and the ``status`` indicates whether the operation is successful and possible failure reason.

Reviewer: ```suggestion
In the ``Response`` message, the ``request_op_code`` is ``START_SCAN``, and the ``status`` indicates whether the operation is successful and the possible failure reason.
```

When the scan is finished, the target will set up a ``Result`` message for each SSID found during the scan, and the information will be in the ``scan_record`` field.

Stop SSID scan
--------------

In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``STOP_SCAN``, over the transport layer.

> The target receives the message, stops WiFi scan, and sets up a ``Response`` message.

Reviewer: ```suggestion
The target receives the message, stops Wi-Fi scan, and sets up a ``Response`` message.
```

> The target receives the message, stops Wi-Fi scan, and sets up a ``Response`` message.

Reviewer: ```suggestion
The target receives the message, stops the Wi-Fi scan, and sets up a ``Response`` message.
```

In the ``Response`` message, the ``request_op_code`` is ``STOP_SCAN``, and the status indicates whether the operation is successful and possible failure reason.

Connect to AP
-------------

In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``SET_CONFIG``, over the transport layer.
The target receives the message, applies the Wi-Fi connection information, and sets up a ``Response`` message.

> In the ``Response`` message, the ``request_op_code`` is ``SET_CONFIG``, and the ``status`` indicates whether the operation is successful and possible failure reason.

Reviewer: ```suggestion
In the ``Response`` message, the ``request_op_code`` is ``SET_CONFIG``, and the ``status`` indicates whether the operation is successful and the possible failure reason.
```

When the connection state changes or an attempt fails, the target will set up a ``Result`` message.
The ``state`` field indicates the current state of the Wi-Fi, and ``reason`` field indicates the failure reason.

Disconnect from AP
------------------

In this workflow, the configurator sends a ``Request`` message, in which the ``op_code`` is set as ``FORGET_CONFIG``, over the transport layer.
The target receives the message, triggers the Wi-Fi disconnection, and sets up a ``Response`` message.

> In the ``Response`` message, the ``request_op_code`` is ``FORGET_CONFIG``, and the ``status`` indicates whether the operation is successful and possible failure reason.

Reviewer: ```suggestion
In the ``Response`` message, the ``request_op_code`` is ``FORGET_CONFIG``, and the ``status`` indicates whether the operation is successful and the possible failure reason.
```

> When the connection state changes or an attempt fails, the target will set up a ``Result`` message, and the ``state`` field indicates the current state of the Wi-Fi, and ``reason`` field indicates the failure reason.

Reviewer: ```suggestion
When the connection state changes or an attempt fails, the target will set up a ``Result`` message, and the ``state`` field indicates the current state of the Wi-Fi, and the ``reason`` field indicates the failure reason.
```

* ``Operation Control Point``: For client to send ``Request`` message to server, and server to send ``Response`` message to client.
* ``Data Out``: For server to send ``Result`` message to the client.

> It takes the functions exposed by Task and event handling part of reading ``Info`` message and  receiving ``Request`` message as the callbacks of corresponding characteristics.
> It provides fucntions for Task and event handling part to send ``Response`` and ``Result`` messages.

Reviewer: ```suggestion
It takes the functions exposed by the Task and event handling part of reading the ``Info`` message and  receiving the ``Request`` message as the callbacks of corresponding characteristics.
It provides functions for Task and event handling part to send ``Response`` and ``Result`` messages.
```

Task and event handling
***********************

> The service uses `nanopb`_ to instantiate the Protocol Buffers-based platform-independent messages in C language.

Reviewer: ```suggestion
The service uses `nanopb`_ to instantiate the Protocol Buffers-based, platform-independent messages in the C language.
```

> It exposes the functions of reading ``Info`` message and  receiving ``Request`` message to transport layer.

Reviewer: ```suggestion
It exposes the functions of reading the ``Info`` message and  receiving the ``Request`` message to the transport layer.
```

> It uses the function of sending ``Response`` and ``Result`` messages provided by transport layer to send these messages.

Reviewer: ```suggestion
It uses the function of sending ``Response`` and ``Result`` messages provided by the transport layer to send these messages.
```

* ``Operation Control Point``: For client to send ``Request`` message to server, and server to send ``Response`` message to client.
* ``Data Out``: For server to send ``Result`` message to the client.

> It takes the functions exposed by the Task and event handling part of reading the ``Info`` message and  receiving the ``Request`` message as the callbacks of corresponding characteristics.

Reviewer: ```suggestion
It takes the functions exposed by the task and event handling part of reading the ``Info`` message and receiving the ``Request`` message as the callbacks of corresponding characteristics.
```

> It provides functions for task and event handling part to send ``Response`` and ``Result`` messages.

Reviewer: ```suggestion
It provides functions for the task and event handling part to send ``Response`` and ``Result`` messages.
```

> It provides functions for Task and event handling part to send ``Response`` and ``Result`` messages.

Reviewer: ```suggestion
It provides functions for task and event handling part to send ``Response`` and ``Result`` messages.
```

Task and event handling
***********************

> The service uses `nanopb`_ to instantiate the Protocol Buffers-based, platform-independent messages in the C language.

Reviewer: ```suggestion
The service uses `nanopb`_ to instantiate the protocol buffers-based, platform-independent messages in the C language.
```
